addons/image/freeimage.c

```diff
@@ -0,0 +1,168 @@
+   ALLEGRO_BITMAP *bitmap = NULL;
+   width = FreeImage_GetWidth(fib);
+   height = FreeImage_GetHeight(fib);
+   bitmap = al_create_bitmap(width, height);
+   if(bitmap) {
+      ALLEGRO_LOCKED_REGION *a_lock = al_lock_bitmap(bitmap,
+         ALLEGRO_PIXEL_FORMAT_ARGB_8888, ALLEGRO_LOCK_WRITEONLY);
+      if (a_lock) {
+         unsigned char *out = (unsigned char *)a_lock->data;
+         int out_inc = a_lock->pitch - (width*4);
+         for(int j=height - 1; j > -1; --j) {
+            for(int i=0; i < width; ++i) {
+               RGBQUAD color;
+               if(FreeImage_GetPixelColor(fib, i, j, &color) == FALSE) {
+                  ALLEGRO_ERROR("Unable to get pixel data at %d,%d\n", i , j);
+               }
+               *out++ = (unsigned char) color.rgbBlue;
+               *out++ = (unsigned char) color.rgbGreen;
+               *out++ = (unsigned char) color.rgbRed;
+               *out++ = (unsigned char) color.rgbReserved;
+            }
+            out += out_inc;
+         }
+         al_unlock_bitmap(bitmap);
+      }
+   }
+   return bitmap;
+}
+
+ALLEGRO_BITMAP *_al_load_fi_bitmap(const char *filename, int flags)
+{
+   FIBITMAP *fib = NULL;
+   ALLEGRO_BITMAP *bitmap = NULL;
+   FREE_IMAGE_FORMAT fif = FIF_UNKNOWN;
+
+   ASSERT(filename);
+   ASSERT(freeimage_initialized == true);
+
+   fif = FreeImage_GetFIFFromFilename(filename);
+   if(fif == FIF_UNKNOWN)
+      fif = FreeImage_GetFileType(filename, 0);
+   if(fif == FIF_UNKNOWN)
+      return bitmap;
+
+   {
+     FIBITMAP *fibRaw = FreeImage_Load(fif, filename, flags);
+     if(!fibRaw)
+       return bitmap;
+     fib = FreeImage_ConvertTo32Bits(fibRaw);
+     FreeImage_Unload(fibRaw);
+     if(!fib)
+       return bitmap;
+   }
+
+   bitmap = _al_fi_to_al_bitmap(fib);
+   FreeImage_Unload(fib);
+   return bitmap;
+}
+
+static unsigned int _fiio_al_read(void *buffer, unsigned size, unsigned count, fi_handle handle) {
+   return (unsigned int) al_fread((ALLEGRO_FILE *)handle, buffer, (count * size));
+}
+
+static unsigned int _fiio_al_write(void *buffer, unsigned size, unsigned count, fi_handle handle) {
+   return (unsigned int) al_fwrite((ALLEGRO_FILE *)handle, buffer, (count * size));
+}
+
+static int _fiio_al_fseek(fi_handle handle, long offset, int origin) {
+   return al_fseek((ALLEGRO_FILE *)handle, offset, origin);
+}
+
+static long _fiio_al_ftell(fi_handle handle) {
+   return (long) al_ftell((ALLEGRO_FILE *)handle);
+}
+
+ALLEGRO_BITMAP *_al_load_fi_bitmap_f(ALLEGRO_FILE *f, int flags)
+{
+   FreeImageIO fio;
+   ALLEGRO_BITMAP *bitmap = NULL;
+   FIBITMAP *fib = NULL;
+   FREE_IMAGE_FORMAT fif = FIF_UNKNOWN;
+
+   ASSERT(f);
+   ASSERT(freeimage_initialized == true);
+
+   fio.read_proc  = _fiio_al_read;
+   fio.write_proc = _fiio_al_write;
+   fio.seek_proc  = _fiio_al_fseek;
+   fio.tell_proc  = _fiio_al_ftell;
+
+   fif = FreeImage_GetFileTypeFromHandle(&fio, (fi_handle)f, 0);
+   if(fif == FIF_UNKNOWN)
+      return bitmap;
```

Comment:
Same as above: return NULL + warning.

Reply:
Fixed.

If there's a FreeImage error, FreeImage should emit a warning from the error handler installed.